src/service/history-impl.js

```diff
@@ -525,6 +545,28 @@ export class HistoryBindingNatural_ {
   }
 
   /**
+   * If this is a hash update the choice of `location.replace` vs
+   * `history.replaceState` is important. Due to bugs, not every browser
+   * triggers `:target` pseudo-class when `replaceState` is called.
+   * See http://www.zachleat.com/web/moving-target/ for more details.
+   * location.replace will trigger a `popstate` event, we temporarily
+   * disable handling it.
+   * @param {string} target
+   *
+   * @override
+   */
+  replaceStateForTarget(target) {
+    this.whenReady_(() => {
+      let hash = target.indexOf('#') == 0 ? target : `#${target}`;
+      this.ignoreUpcomingPopstate_ = true;
+      // TODO(mkhatib, #6095): Chrome iOS will add extra states for location.replace.
+      this.win.location.replace(hash);
+      this.ignoreUpcomingPopstate_ = false;
+      this.historyReplaceState_();
```

Comment:
`#replaceState` should trigger a `popstate`, too right? Well, call it first, that's your "real" `popstate`. On every `popstate`, record the current URL. Now, we need to call `#replace` to get `:target` support, but it'll trigger it's own `popstate` (a duplicate). But because we know what URL we were at, we know this is a duplicate URL. Don't record this `popstate`.